### PR TITLE
QA: Implement Avatar builder Frontend (Fixes #813)

### LINKED ISSUE #813: As a user, I want to build an Avatar, so I can show off my bad self.
### User Story

[comment]: # (As a <user type>, I want to <task> so that <goal>.)
As a user, I want to build an Avatar, so I can show off my bad self.

### Why Is this Needed?

To make profiles on Gitcoin unique and peronsable... without being creepy

### Description

In the new FTUX flow ( https://github.com/gitcoinco/web/issues/441 ) please add a page that is an avatar builder.  This avatar builder will allow a user to select facial features (hats, glasses, earrings, beards, beard alt, mustaches, skin color, head shape, ears torso, mouth, head, nose, eyes, hair, hair color, and build an avatar

#### Grid
screen shot 2018-04-05 at 3 03 51 pm

#### Library of Components
screen shot 2018-04-05 at 3 04 23 pm

screen shot 2018-04-05 at 3 04 57 pm
screen shot 2018-04-05 at 3 04 52 pm
screen shot 2018-04-05 at 3 04 47 pm

#### Final result

Please disregard the 3 icons overlaying the avatar. 

screen shot 2018-04-05 at 3 05 05 pm

### Avatar builder specifications

screen shot 2018-04-07 at 12 49 20 pm

screen shot 2018-04-07 at 12 50 24 pm

screen shot 2018-04-07 at 12 50 34 pm

screen shot 2018-04-07 at 12 50 43 pm

### Current Behavior
Users Github Profile image is Used

### Expected Behavior
* The avatar builder should show a live preview of the avatar as it's being assembled.
* The avatar builder should save and automatically sync to user's Gitcoin profile.
* The avatar builder should allow the user to save their avatar to their gitcoin profile

### Definition of Done
- All functionality exists and works as outlined in mockup and guidelines.
- Visual designs leverage Gitcoin [UI Kit styling](https://github.com/gitcoinco/creative/tree/master/Product/00_UI_Kit) (fonts, colors, etc) found in the Gitcoin Creative Repo.
- **Characteristics List** On the left show as square dots until user has made a selection of corresponding characteristic on the right, once selected, dot turns into a checkmark.
- **Avatar Work Area**  Please follow interaction hover and click states as shown in the visual guide
- Color bars will need to be reflective of the section the user is in. For instance skin tone colors will be different than accessory colors.
- A user will only be able to select one option from the following sections:
Face, Eyes, Nose, Mouth, Ears, Hair, Facial Hair, Clothing, and Background.
- A user may select multiple options from the following sections: Accessories.
- Once user completes the avatar, a light confirmation should display at the upper right hand corner of the avatar builder (placeholder is shown in the mockup).

### Assets
* [Avatar Builder Sketch File](https://github.com/gitcoinco/creative/tree/master/Product/Avatar)
* [Avatar Art Assets](https://github.com/gitcoinco/creative/tree/master/Product)
* [Gitcoin UI Kit ]((https://github.com/gitcoinco/creative/tree/master/Product/00_UI_Kit) 

### Additional Information
[comment]: # (Any other information that would be useful, content, screenshots, etc.)
Other Avatar Projects
* http://www.hexatar.com/
* https://github.com/alizohaib/avatar-maker

### Reviewers
@pixelantdesign @owocki

### PR DESCRIPTION
**Pull request by Smona**
##### Description
Adds an avatar-builder to be used in the new FTUX flow. 

##### Checklist

- [x] linter status: 100% pass
- [x] changes don't break existing behavior
- [x] commit message follows [commit guidelines](https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#step-4-commit)

##### Affected core subsystem(s)
ui

##### Testing
Tested in FF, Safari, and Chrome

##### Refers/Fixes
Fixes: #813

### CODE REVIEW

**Comment by thelostone-mc** [comment]
The number of svgs in this PR is too damn high 😂 
Code from what I've seen makes sense  :D 

@Smona Oh just a heads up run the linters once you finish up your work ^_^
`npm run eslint:fix` and `npm run stylelint:fix` so that the CI build passes

File: app/assets/v2/images/avatar/Accessories/Earring-0.svg [hunk 1]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Generator: Adobe Illustrator 22.1.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
```
Comment (mbeacom) [inline app/assets/v2/images/avatar/Accessories/Earring-0.svg]: We'll want to run `svgo` against all of the newly added SVGs before this gets merged.
Instructions here: https://github.com/gitcoinco/web/blob/master/docs/CONTRIBUTING.md#contributing-static-assets

File: app/assets/v2/css/avatar_builder.css [hunk 2]
```diff
@@ -0,0 +1,120 @@
+body {
+  padding: 35px;
+  max-width: 1000px;
+  margin: 0 auto;
+}
```
Comment (mbeacom) [inline app/assets/v2/css/avatar_builder.css]: You'll want to run `make fix` in the repo directory.

File: app/app/urls.py [hunk 3]
```diff
@@ -148,6 +148,7 @@
     url(r'^medium/?', retail.views.medium, name='medium'),
     url(r'^github/?', retail.views.github, name='github'),
     url(r'^youtube/?', retail.views.youtube, name='youtube'),
+    url(r'^avatar/?', retail.views.avatar),
```
Comment (mbeacom) [inline app/app/urls.py]: Please use `re_path()` for regex urls and `path()` for standard versus `url()`

File: app/assets/v2/images/avatar/Accessories/Earring-0.svg (hunk 1 above)
Comment (Smona) [reply app/assets/v2/images/avatar/Accessories/Earring-0.svg]: will do once confirmed all svgs are added

**Comment by owocki** [comment]
just tested this.. very exciting progress!

some notes:
* i wonder if we can put it on a page with a more standard header / nav
* is there a way to remove beards, earrings, hats, etc?
* 'save avatar' doesnt seem to work right now.

**Comment by Smona** [comment]
@owocki i'm leaving the headr/nav/footer includes and save/do later button actions for the integration step with @thelostone-mc. Ability to remove optional categories is next on my todo list.

**Comment by thelostone-mc** [comment]
@Smona So we've got the onboarding code into the repo -> `onboarding.html`
If you check that out -> you'll notice the final step is the avatar  where an image stub has been added [here](https://github.com/gitcoinco/web/blob/master/app/dashboard/templates/onboard.html#L171)

Could we link your html into this ? (not port over , just link it cause avatar builder is big and might get reused in the other onboarding flows -> might make sense to let it exist as it's independent component )

We'll have to add the `Skip` button along with `Previous`  / `Done` which is simple

File: app/dashboard/templates/onboard.html [hunk 4]
```diff
@@ -22,6 +22,17 @@
...
+    <link rel="stylesheet" href="{% static "v2/css/avatar_builder.css" %}" />
+    <style>
+      #avatar-builder__sections li.complete {
+        list-style-image: url('{% static "v2/images/check-dark.svg" %}');
+      }
+      .avatar-option.selected::after {
+        content: url('{% static "v2/images/check.svg" %}');
+        background-color: #150069;
+        line-height: 20px;
+      }
```
[the remark this answers is not in the document]
Comment (Smona) [reply app/dashboard/templates/onboard.html]: Done in 1bc4d

**Comment by thelostone-mc** [comment]
@Smona You'll probably hate me for this -> but we've started using `rem` as opposed to `px` -> could you update the CSS bit to that ?

(If this is tech creep, I'll shoot a follow up PR and update it myself - no worries ^_^ )

**Comment by Smona** [comment]
@owocki the frontend is tested and meets the spec. Since @mbeacom is working on the avatar backend, I think that means this is good to merge pending your review

**Comment by owocki** [comment]
got it; @mbeacom how do you want to do the PRs for the frontend vs backend?

**Comment by owocki** [comment]
1. is this ready for QA yet?
2. might be good to resolve those merge conflicts

**Comment by mbeacom** [comment]
@owocki 

1. Not quite yet.  I'll be committing up my BE changes today.  This includes (but is not limited to!):

    - saving final avatar asset in S3
    - associating avatar with Profile
    - Optional promotion of GC avatar to current profile avatar
    - dynamic templating for SVG base assets for easy expansion of color options for any avatar item
    - stored avatar configurations

2. Resolving merge conflicts now!

**Comment by owocki** [comment]
just tested... a lot of it looks good.  there are some things to fix before this is ready to go live:.

1. `onboard/contributor` is not linked anywhere
2. `save avatar` does not work on `onboard/contributor/avatar` (spoke with mark, it sounds like it might be because my local s3 is not setup?)
3. i think [this interface](http://bits.owocki.com/380g1K1X0J1G/Screen%20Recording%202018-06-15%20at%2001.29%20PM.gif) is confusing.  is everything in green my 'saved skills'?  why are they listed in two difference places? 
4. (3 contd) experience level -- experience level with what?  programming?  with all of the above skills?
5. after i go through the onboard flow im taken to [this page](http://bits.owocki.com/1X0k2t1i1V2b/Screen%20Shot%202018-06-15%20at%201.31.32%20PM.png) which has the folowing issues: (1) helmet never stops loading (2) 'getting started guide' (both in the header and footer) and 'more contributor tips' do not link anywhere.  (3) 'got it' button doesnt do anything except refresh the page and take me to the same broken page. (4) [js errors](http://bits.owocki.com/2e0z0g3M0T2B/Screen%20Shot%202018-06-15%20at%201.32.56%20PM.png)

@PixelantDesign @Smona @thelostone-mc @SaptakS @mbeacom

**Comment by PixelantDesign** [comment]
@thelostone-mc 

> The getting starting guide and more contributor tips are yet to be linked (I'll add it once @PixelantDesign gives me the links)

The guide should link to developer guide. Is that what you need? We could probably remove the link to more tips on the right side for now.

When we have the How it works built - I could see it being useful here as well.

@owocki @thelostone-mc 
The tag screen is not a new paradigm. It's used in a few places on Google.

Basically - A user selects the top items that they see which are relevant to their expertise (Aditya mentioned that Mark is going to make this area dynamic - this seems to be one of the outstanding items on the ftux flow). Any other relevant skills that the user doesn't see see on the top they'd add below just by typing in. Below that they'd add experience level. Pretty straight forward. Let me know which part is unclear.

Could someone deploy on staging so that I can review?

**Comment by owocki** [comment]
[this button](http://bits.owocki.com/1S1V1p1Z3N0u/Screen%20Shot%202018-06-18%20at%209.46.44%20AM.png) looks jacked

**Comment by owocki** [comment]
two more small bugs:
1. when you select white as your skin color, and then add ears, they are the wrong color
2. the baseball caps all seem like they need to be moved up like 5%

**Comment by owocki** [comment]
more bugs:
1. 'do this later' button on the avatar doesnt do anything
2. 'save avatar' button 404s http://bits.owocki.com/3p0e3s3Q0f2N/Screen%20Recording%202018-06-18%20at%2009.48%20AM.gif

**Comment by owocki** [comment]
> As per the design put up by Alisa -> this is how it work

OK i defer to Alisa on this one.

> > (3 contd) experience level -- experience level with what? programming? with all of the above skills?

> Yup!

Can we make this clearer in the UI?

> The getting starting guide and more contributor tips are yet to be linked (I'll add it once @PixelantDesign gives me the links)


@pixelantdesign can you hook it up?  edit: nevermind, looks like she already responded..  you can find the links to these guides in /help

> helmet never stopped loading probably cause the API call failed( remember the insane API call load time I was talking about ? )

in the console, the API call is never even made.  to reproduce this, go through the entire flow without logging in and land your self on the /explorer page

===============================================

guys, if you just go through the entire flow and just try a few different permutations (diff screen sizes, logged in vs logged out, etc) youll find a lot of outstanding bugs.  there's nothing special i'm doing here other than just using the product like a bunch of different users would.

**Comment by owocki** [comment]
a few other things to firm up 

1.  while i can now save my avatar… if i reload the page and want to *edit* the avatar i ahve to start over again
2. the UI doesnt do anything to give me confidence that my avatar has saved when it does successfully save

**Comment by owocki** [comment]
two more bugs:

1. if i create an avatar, and refresh the page to create another avatar and then save that second avatar, then go to stage.gitcoin.co -- the avatar in the top right has not updated.
2. my avatar never seem to update at https://stage.gitcoin.co/profile/owocki

**Comment by SaptakS** [comment]
- [x] Buttons have been fixed
- [x] when you select white as your skin color, and then add ears, they are the wrong color - Fixed
- [x] Show visual indication of saving.

**Comment by SaptakS** [comment]
> I agree with Kevin's point about immediately taking the user to the next option once they've customized a feature on the avatar

I don't totally agree with this. I agree that maybe there should be some way to take to the next step without having to click through the links, but taking user to next option immediately after selecting isn't right.
 So let's say I just selected a head, now I can't be taken immediately to the next option since I might decide to play around with other shapes and colors.
Now if I want to take to next step after playing around is done, then we need some kind of button or something to give the indication that customisation of this feature is done and I am ready to go to next step.

But I feel there is already too many buttons in that page.

**Comment by owocki** [comment]
> I don't totally agree with this. I agree that maybe there should be some way to take to the next step without having to click through the links, but taking user to next option immediately after selecting isn't right.

maybe just using javascript to allow the user to navigate through the avatar builder with a left and right keys?  and then adding an info button that tells them about the keyboard shortcuts?

**Comment by SaptakS** [comment]
> while i can now save my avatar… if i reload the page and want to edit the avatar i ahve to start over again

Done

**Comment by mbeacom** [comment]
Things to do before merging (**PLEASE DO NOT MERGE UNTIL COMPLETED**):

- [ ] Delete all new migrations and create them as one migration (this is to avoid the necessity to fake migrations on staging while we QA as we iterate through the reqs to deploy)
- [ ] Run `make fix` and resolve any errors/warnings.
- [ ] Replace any newly introduced `print()` statements with `logger` statements.
- [ ] Final QA on latest commit against local (to ensure all newly added functionality works for community contributors)
- [ ] Final QA on latest commit against staging (to ensure all prod/staging specific logic responds appropriately).

**Comment by PixelantDesign** [comment]
Hi @thelostone-mc @SaptakS @mbeacom 
I noticed a few more things the avatar builder and would also like us to look at editing some swatch colors.

- [ ] Currently there are 4 buttons shown on staging, please update to 3 buttons as shown on the screenshot below.
- [ ] Please remove drop shadow on buttons.
- [ ] Please make buttons the same size and consistent with the rest of the ftux flow.
- [ ] Color - Please Add #EEE3C1  AFTER  #F8D5C2
- [ ] Color - Please remove Remove #3F2A18 from the skin palette (color on the far right) since the facial features don't show on this color.
- [ ] Do this later button currently doesn't work. Clicking do this later should advance the user to the next step - skill selection.

screen shot 2018-06-19 at 10 00 39 pm

I'll save the rest of the ftux comments for the next PR. Thanks!

**Comment by Smona** [comment]
So exciting to see this go live! :smiley:

**Comment by thelostone-mc** [comment]
@Smona it is live ^_^
URL : `gitcoin.co/onboard/contributor`

**Pull request merged**